### PR TITLE
API level component tests

### PR DESCRIPTION
**Pull request by BarisSari**
I haven't used python-bdd because I wanted to keep things simple for these tests. But, I might change them when I'm working on integration tests in the future.

### CODE REVIEW

**Comment by BarisSari** [comment]
When we merge this PR, I'll implement these kind of tests for other microservices, too.

File: app/api/enrolments.py [hunk 1]
```diff
@@ -3,10 +3,10 @@
 from app.repositories.s3_callback_repo import S3CallbackRepo
 from app.repositories.s3_enrolment_repo import S3EnrolmentRepo
 from app.requests.enrolment_requests import NewEnrolmentRequest
-from app.use_cases.create_new_enrolment import CreateNewEnrolment
-from app.use_cases.get_callbacks_list import GetCallbacksList
-from app.use_cases.get_enrolment import GetEnrolmentByID
-from app.use_cases.get_enrolment_status import GetEnrolmentStatus
+from app.use_cases import create_new_enrolment as cne
```
Comment (TheAyeshaSiddiqa) [inline app/api/enrolments.py:6]: I think you Create enrolment shouldn't be there in employer-callback. It should only be in employer-admin.

File: app/api/enrolments.py [hunk 2]
```diff
@@ -3,10 +3,10 @@
...
+from app.use_cases import get_callbacks_list as gcl
```
Comment (TheAyeshaSiddiqa) [inline app/api/enrolments.py:7]: Same as above. It should be in employer-admin, not here.

File: app/api/enrolments.py [hunk 3]
```diff
@@ -3,10 +3,10 @@
...
+from app.use_cases import get_enrolment as ge
+from app.use_cases import get_enrolment_status as ges
```
Comment (TheAyeshaSiddiqa) [inline app/api/enrolments.py:9]: Same as above. It should be in employer-admin, not here.

File: app/repositories/s3_enrolment_repo.py [hunk 4]
```diff
@@ -15,63 +17,116 @@ class S3EnrolmentRepo(EnrolmentRepo):
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        with handle_s3_errors():
-            self.s3 = boto3.client("s3", **settings.s3_configuration)
-
-    def save_enrolment(self, enrollment: dict):
+        self.s3 = boto3.client("s3", **settings.s3_configuration)
```
Comment (TheAyeshaSiddiqa) [inline app/repositories/s3_enrolment_repo.py:20]: add handle_s3_errors() over it

File: app/use_cases/get_enrolment.py [hunk 5]
```diff
@@ -12,8 +13,17 @@ class Config:
...
+            # check if enrolment exists or not
+            if not self.enrolment_repo.enrolment_exists(
+                enrolment_id, bucket=settings.ENROLMENT_BUCKET
+            ):
+                return ResponseFailure.build_from_validation_error(
+                    message="enrolment_id " + enrolment_id + " is not valid."
+                )
             enrolment = self.enrolment_repo.get_enrolment(enrolment_id=enrolment_id)
-        except Exception as e:  # noqa - TODO: handle specific failure types
+            code = SuccessType.SUCCESS
```
Comment (TheAyeshaSiddiqa) [inline app/use_cases/get_enrolment.py:24]: its not needed, success is by default.

File: app/use_cases/get_enrolment.py [hunk 6]
```diff
@@ -12,8 +13,17 @@ class Config:
...
+            message = "Enrolment has been fetched."
+        except Exception as e:
             return ResponseFailure.build_from_resource_error(message=e)
 
-        return ResponseSuccess(value=enrolment)
+        return ResponseSuccess(value=enrolment, message=message, type=code)
```
Comment (TheAyeshaSiddiqa) [inline app/use_cases/get_enrolment.py:29]: its not needed, success is by default.

File: tests/api/test_callbacks.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+from app.main import app
+from app.responses import FailureType, ResponseFailure, ResponseSuccess, SuccessType
+from tests.test_data.callback_provider import CallbackDataProvider
+
+test_data = CallbackDataProvider()
+client = TestClient(app)
+
+
+@mock.patch("app.use_cases.create_new_callback.CreateNewCallback")
+def test_create_callbacks_success_created(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_callbacks.py:15]: test_create_callback*

File: tests/api/test_callbacks.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    callback = Callback(**json_result.get("value"))
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert callback == test_data.sample_callback
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.create_new_callback.CreateNewCallback")
+def test_create_callbacks_success(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_callbacks.py:36]: test_create_callback*

File: tests/api/test_callbacks.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    callback = Callback(**json_result.get("value"))
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.SUCCESS.value
+    assert callback == test_data.sample_callback
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.create_new_callback.CreateNewCallback")
+def test_create_callbacks_failure(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_callbacks.py:55]: test_create_callback*

File: tests/api/test_enrolments.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+from fastapi.testclient import TestClient
+
+from app.domain.entities.enrolment import Enrolment
+from app.main import app
+from app.responses import FailureType, ResponseFailure, ResponseSuccess, SuccessType
+from tests.test_data.callback_provider import CallbackDataProvider
+from tests.test_data.enrolment_data_provider import DataProvider
+
+test_data = CallbackDataProvider()
+test_data_2 = DataProvider()
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:12]: maybe call it EnrolmentDataProvider, for specification.

File: tests/api/test_enrolments.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert enrolment == test_data.sample_enrolment
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.create_new_enrolment.CreateNewEnrolment")
+def test_create_enrolment_success(use_case):
+    message = "The callback has been fetched from the server."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:39]: The enrolment has been fetched from the server.*

File: tests/api/test_enrolments.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    enrolment = Enrolment(**json_result.get("value"))
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert enrolment == test_data.sample_enrolment
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.create_new_enrolment.CreateNewEnrolment")
+def test_create_enrolment_success(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:38]: There is no code in use_cases.create_new_enrolment.CreateNewEnrolment for fetching enrolment. there is only for creating a one.

File: tests/api/test_enrolments.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    response = client.post("/enrolments", data=data.json())
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_enrolment.GetEnrolmentByID")
+def test_get_enrolment_by_id_success_created(use_case):
+    code = SuccessType.CREATED
+    message = "The enrolment has been created."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:73]: its "Enrolment has been fetched." in the code.

File: tests/api/test_enrolments.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+
+    data = test_data.sample_enrolment_request
+    response = client.post("/enrolments", data=data.json())
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_enrolment.GetEnrolmentByID")
+def test_get_enrolment_by_id_success_created(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:71]: GetEnrolmentByID doesn't create an enrolment. why are making created test?

File: tests/api/test_enrolments.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert enrolment == test_data.sample_enrolment
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.get_enrolment.GetEnrolmentByID")
+def test_get_enrolment_by_id_success(use_case):
+    message = "The callback has been fetched from the server."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:93]: its "Enrolment has been fetched." in the code.

File: tests/api/test_enrolments.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    response = client.get(f"/enrolments/{test_data.enrolment_id}")
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_enrolment_status.GetEnrolmentStatus")
+def test_get_enrolment_status_success_created(use_case):
+    code = SuccessType.CREATED
+    message = "The enrolment has been created."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:126]: "Enrolment status has been fetched."*

File: tests/api/test_enrolments.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+    )
+
+    response = client.get(f"/enrolments/{test_data.enrolment_id}")
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_enrolment_status.GetEnrolmentStatus")
+def test_get_enrolment_status_success_created(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:124]: GetEnrolmentStatus doesn't create an enrolment.

File: tests/api/test_enrolments.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert enrolment == test_data.sample_enrolment
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.get_enrolment_status.GetEnrolmentStatus")
+def test_get_enrolment_status_success(use_case):
+    message = "The callback has been fetched from the server."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:146]: "Enrolment status has been fetched."*

File: tests/api/test_enrolments.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+    response = client.get(f"/enrolments/{test_data.enrolment_id}/status")
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_callbacks_list.GetCallbacksList")
+def test_get_callbacks_list_for_enrolment_success_created(use_case):
+    code = SuccessType.CREATED
+    message = "The enrolment has been created."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:179]: "Callbacks list has been fetched"

File: tests/api/test_enrolments.py [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+
+    use_case().execute.assert_called_with(data)
+    assert response.status_code == SuccessType.CREATED.value
+    assert json_result.get("value") == test_data_2.callbacks_list1_json
+    assert json_result.get("message") == message
+
+
+@mock.patch("app.use_cases.get_callbacks_list.GetCallbacksList")
+def test_get_callbacks_list_for_enrolment_success(use_case):
+    message = "The callback has been fetched from the server."
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:198]: "Callbacks list has been fetched"*

File: tests/api/test_enrolments.py [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+    )
+
+    response = client.get(f"/enrolments/{test_data.enrolment_id}/status")
+
+    assert response.status_code == FailureType.RESOURCE_ERROR.value
+    assert response.json() == {"detail": message}
+
+
+@mock.patch("app.use_cases.get_callbacks_list.GetCallbacksList")
+def test_get_callbacks_list_for_enrolment_success_created(use_case):
```
Comment (TheAyeshaSiddiqa) [inline tests/api/test_enrolments.py:177]: GetCallbacksList doesn't create anything

File: tests/repositories/test_s3_enrolment_repo.py [hunk 22]
```diff
@@ -24,110 +24,284 @@ def test_s3_initialisation(boto_client):
...
-def test_save_enrolment(boto_client, uuid4):
-    """
-    Ensure the S3Enrolmentrepo returns an object with OK data
-    and that an appropriate boto3 put call was made.
-    """
-    fixed_uuid_str = "1dad3dd8-af28-4e61-ae23-4c93a456d10e"
-    uuid4.return_value = UUID(fixed_uuid_str)
+@mock.patch("app.utils.random.Random.get_uuid")
+@mock.patch("boto3.client")
+def test_create_enrolment(boto_client, get_uuid):
```
Comment (TheAyeshaSiddiqa) [inline tests/repositories/test_s3_enrolment_repo.py:29]: I think you can remove this test as well. As it should be in employer admin

File: tests/repositories/test_s3_enrolment_repo.py [hunk 23]
```diff
@@ -24,110 +24,284 @@ def test_s3_initialisation(boto_client):
...
-@patch("uuid.uuid4")
-@patch("boto3.client")
-def test_get_enrolment_status(boto_client, uuid4, callback_repo_list):
-    """
-    Ensure the S3Enrolmentrepo returns an object with OK data
-    and that get_enrolment_status returns appropriate object.
-    """
+@mock.patch("json.loads")
+@mock.patch("boto3.client")
+def test_get_callbacks_list(boto_client, json_loads):
```
Comment (TheAyeshaSiddiqa) [inline tests/repositories/test_s3_enrolment_repo.py:71]: I think you can remove this test as well. As it should be in employer admin

File: tests/repositories/test_s3_enrolment_repo.py [hunk 24]
```diff
@@ -24,110 +24,284 @@ def test_s3_initialisation(boto_client):
...
+    boto_client.return_value.list_objects = list_objects_sample_content_empty
+    json_loads.return_value = callback
+
+    callbacks_list = repo.get_callbacks_list(enrolment_id)
+
+    assert callbacks_list == {"callbacks_list": []}
+
+
+@mock.patch("app.repositories.s3_enrolment_repo.S3EnrolmentRepo.get_callbacks_list")
+def test_get_enrolment_status(get_callbacks_list):
```
Comment (TheAyeshaSiddiqa) [inline tests/repositories/test_s3_enrolment_repo.py:108]: I think you can remove this test as well. As it should be in employer admin

File: tests/test_data/callback_provider.py [hunk 25]
```diff
@@ -3,6 +3,7 @@
 from app.domain.entities.callback import Attachment, Callback
 from app.domain.entities.enrolment import Enrolment
 from app.requests.callback_requests import CallbackRequest
+from app.requests.enrolment_requests import NewEnrolmentRequest
 
 
 class CallbackDataProvider:  # (BaseModel):
```
Comment (TheAyeshaSiddiqa) [inline tests/test_data/callback_provider.py:9]: rename filename: callback_provider.py to callback_data_provider.py

File: app/api/callbacks.py [hunk 26]
```diff
@@ -4,7 +4,7 @@
 from app.repositories.s3_enrolment_repo import S3EnrolmentRepo
 from app.requests.callback_requests import CallbackRequest
 from app.responses import SuccessType
-from app.use_cases.create_new_callback import CreateNewCallback
+from app.use_cases import create_new_callback as uc
```
Comment (TheAyeshaSiddiqa) [inline app/api/callbacks.py:7]: why do you think this approach is better?

File: app/repositories/s3_enrolment_repo.py [hunk 27]
```diff
@@ -15,63 +17,116 @@ class S3EnrolmentRepo(EnrolmentRepo):
...
 
-        enrl = Enrolment(**enrollment)
-        ref_hash = Random.get_str_hash(enrl.internal_reference)
+    def create_enrolment(self, internal_reference: str) -> Enrolment:
```
Comment (TheAyeshaSiddiqa) [inline app/repositories/s3_enrolment_repo.py:22]: you may remove this method

File: app/repositories/s3_enrolment_repo.py [hunk 28]
```diff
@@ -15,63 +17,116 @@ class S3EnrolmentRepo(EnrolmentRepo):
...
                 Bucket=settings.ENROLMENT_BUCKET,
             )
+            return False
         except Exception:
             return True
-        else:
-            return False
 
-    def get_enrolment(self, enrolment_id: str):
+    def get_callbacks_list(self, enrolment_id: str):
```
Comment (TheAyeshaSiddiqa) [inline app/repositories/s3_enrolment_repo.py:85]: you may remove this method

File: app/repositories/s3_enrolment_repo.py [hunk 29]
```diff
@@ -15,63 +17,116 @@ class S3EnrolmentRepo(EnrolmentRepo):
...
+                    Key=row["Key"], Bucket=settings.CALLBACK_BUCKET
+                )
+            callback = Callback(**json.loads(obj["Body"].read().decode()))
+            callbacks_list.append(
+                {"callback_id": callback.callback_id, "received": callback.received}
+            )
+        return {"callbacks_list": callbacks_list}
 
-    def get_enrolment_status(self, enrolment_id: str, callbacks_list: list):
+    def get_enrolment_status(self, enrolment_id: str):
```
Comment (TheAyeshaSiddiqa) [inline app/repositories/s3_enrolment_repo.py:102]: you may remove this method

**Review by TheAyeshaSiddiqa — CHANGES_REQUESTED** [review]
some minor changes. as no other projects are synchronized. So you may remove all the code for employer-admin from here. that is not needed by employer-callback.

I'll suggest to make the changes first (wrong messages returned) before removing the code. 
So these api tests could be copied to employer-admin.

File: app/api/callbacks.py (hunk 26 above)
Comment (BarisSari) [reply app/api/callbacks.py:7]: I made this change to mock the usecase. Otherwise, it wasn't possible to do it.

**Comment by BarisSari** [comment]
Thanks for the review, @TheAyeshaSiddiqa. I removed everything related to Enrolment except `S3EnrolmentRepo.get_enrolment()`.

**Comment by TheAyeshaSiddiqa** [comment]
> Thanks for the review, @TheAyeshaSiddiqa. I removed everything related to Enrolment except `S3EnrolmentRepo.get_enrolment()`.

Have you moved the code to employer-admin before removing?

**Comment by BarisSari** [comment]
> Have you moved the code to employer-admin before removing?

No, I haven't moved it because I have already used the code from **employer-admin**. So, I think employer-admin repository is more updated than the code in this repository.

**Comment by TheAyeshaSiddiqa** [comment]
> > Have you moved the code to employer-admin before removing?
> 
> No, I haven't moved it because I have already used the code from **employer-admin**. So, I think employer-admin repository is more updated than the code in this repository.

I was asking about API tests

**Comment by BarisSari** [comment]
> I was asking about API tests

Aah, no. Not yet but I'm planning to move it later today.

File: tests/test_data/callback_data_provider.py [hunk 30]
```diff
@@ -110,3 +111,6 @@ def __init__(self):
             shared_secret=self.shared_secret,
             internal_reference=self.internal_reference,
         )
+        self.sample_enrolment_request = NewEnrolmentRequest(
```
Comment (TheAyeshaSiddiqa) [inline tests/test_data/callback_data_provider.py:114]: I think NewEnrolmentRequest is also not needed here

**Review by TheAyeshaSiddiqa — CHANGES_REQUESTED** [review]

File: tests/test_data/callback_data_provider.py (hunk 30 above)
Comment (BarisSari) [reply tests/test_data/callback_data_provider.py:114]: Yes, that's right. I'll check the data provider and remove unnecessary stuff.

**Review by TheAyeshaSiddiqa — APPROVED** [review]

**Pull request merged**